Implement inflation-related functions

AaronVanGeffen opened the pull request:
This is working towards implementing #361.

AaronVanGeffen commented on src/openloco/scenario.cpp:
```diff
@@ -1,12 +1,55 @@
 namespace openloco::scenario
 {
+    // Inflation rates?
+    static uint32_t _4FDEC0[32] = {
```
```suggestion
    static const uint32_t _4FDEC0[32] = {
```

AaronVanGeffen commented on src/openloco/scenario.cpp:
```diff
@@ -23,4 +66,95 @@ namespace openloco::scenario
+            _525EDE[i] = 0;
+        }
+
+        uint32_t ecx = year - 1900;
```
If we change this year to below 1900, the game's balance will be thrown off — everything will become more expensive due to 100 extra years worth of inflation.

Perhaps we should branch off here and not compute any inflation prior to 1900, for now?

AaronVanGeffen replied:
I've implemented the proposed solution, for now.

AaronVanGeffen commented:
@kr4g3n This should be it. If you have time, please check out [the AppVeyor CI build](https://ci.appveyor.com/project/AaronVanGeffen/openloco/builds/25205382/artifacts).

kr4g3n commented:
Thank you. I will try it tomorrow (left my laptop at home today)

kr4g3n commented:
So I ran this version using Thrillville109's Locomtion Expanded Mod to test the basics.
Without the patch installed, and starting one of Thrillville109's scenarios (start year 1833) I had a starting fund of $18,000,000 and track cost around $125,000.


With the patch installed, I started off with $10,000 and track was around $125/section to lay.

Thrillville109's mod has road vehicles that start in 18000, but the first train appears in 1825. However, [Build Track] option is not disabled. I tried to build track in 1823 and the game crashed to desktop, building in 1825 (after trains appeared) which laid tracks just fine.

So I guess if there was a way to add a control that disable's track laying until a train is discovered, that would be great. Obviously, the first train year would be a variable, based on other's mods,

AaronVanGeffen commented:
Thanks for testing it out. I wonder what causes it to crash, though — I'm not able to reproduce it.

![Screenshot_20190612_111431](https://user-images.githubusercontent.com/604665/59338767-4abf9a80-8d03-11e9-810c-57712d53065c.png)

I'll it again once I've gotten hold of the mod you've been using.

kr4g3n commented:
It might be an issue in that mod then, I could never even get it to start on my steam downloaded version (would always freeze at generating terrain) but the mod loads fine in openloco. I'll try and get you a screenshot.

kr4g3n commented:
I just tried it on the laptop I take to work, no issues. Trying again on my Alienware M18Xr2 incase it's machine dependant...

kr4g3n commented:
OK tried on my Alienware - first I retried the scenario I built last night. SAME THING. 

This is the scenario I built. Just in case you can find anything  in there? I set it to flat and changed the sea level, hill density, no of forests, no of towns, and number of industries.
[TrainTable.zip](https://github.com/OpenLoco/OpenLoco/files/3281775/TrainTable.zip)

I also seleceted ALL vehicles and Airports from thrillville's mod.
[LocomotionExpanded.zip](https://github.com/OpenLoco/OpenLoco/files/3281781/LocomotionExpanded.zip)

I weill have to do some more testing later

kr4g3n commented:
First, sorry, but I uploaded the wrong scenario file earlier, I'm at work right now so can't upload the other one.

I have however tried some more tests using the boulder breakers map, each time I was able to place tracks etc, but it randomly crashes to desktop. 

Is there a dump file I can send your way? 

Both my machines are Win10

In addition, I noticed that if I start any date other than 1800, I don't actually get that start date. 

Example one game I started in 1823, but the actual start year was 1843.
And if I tried to start in 1825, it started in 1849.

marijnvdwerf commented on src/openloco/scenario.cpp:
```diff
@@ -23,4 +66,97 @@ namespace openloco::scenario
+    // NB: called in sub_46E2C0 below, as well in openloco::date_tick.
+    void sub_46E239()
+    {
+        for (uint8_t i = 0; i < 32; i++)
```
I think the entire highlighted area is useless, as `_525EDE` is only ever filled with zeroes.
<img width="420" alt="loc_46e238" src="https://user-images.githubusercontent.com/737603/59449489-7a57cb00-8e07-11e9-98c2-dd31bb5aac8d.png">

AaronVanGeffen replied:
What makes you sure it's zeros? I think it represents the highest 16 bits of a `money48` struct.

marijnvdwerf replied:
As I said, `525EDE` is only ever filled with zeroes. Just look at the references.

duncanspumpkin replied:
You can see what CS tried to do but he performed the calc wrong. The lower bits from the second shrd were meant to be put into 525EDE but instead are discarded. So 525EDE does indeed do nothing.

marijnvdwerf commented on src/openloco/scenario.cpp:
```diff
@@ -1,12 +1,55 @@
+
+    static loco_global<uint32_t[32], 0x00525E5E> currencyMultiplicationFactor;
+
+    // NB: _525EDE[2] is dword_525E66, referred to by four subs for (price) computation.
```
Comment is incorrect, as it applies to the previous variable instead

marijnvdwerf commented on src/openloco/scenario.cpp:
```diff
@@ -23,4 +66,97 @@ namespace openloco::scenario
+
+        // OpenLoco allows 1800 as the minimum year, whereas Locomotion uses 1900.
+        // Treat years before 1900 as though they were 1900 to not change vanilla scenarios.
+        uint32_t ecx = std::max<uint32_t>(year, 1900) - 1900;
```
something like `std::clamp(year, 1900, 2030) - 1900` would be easier to read.

marijnvdwerf commented on src/openloco/scenario.cpp:
```diff
@@ -23,4 +66,97 @@ namespace openloco::scenario
+        _525FB5 = 1;
+
+        // NB: only called here.
+        call(0x0042FDE2);
```
Goes through all companies and sets the available vehicles.

marijnvdwerf commented on src/openloco/scenario.cpp:
```diff
@@ -1,12 +1,55 @@
 namespace openloco::scenario
 {
+    // Inflation rates?
+    static const uint32_t _4FDEC0[32] = {
```
List of fixed-point percentages. (12 bits for fractions)

marijnvdwerf commented on src/openloco/scenario.cpp:
```diff
@@ -1,12 +1,55 @@
+        20,
+    };
+
+    static loco_global<uint32_t[32], 0x00525E5E> currencyMultiplicationFactor;
```
Assuming these values start at `1`, this is fixed point with 10 fractional bits.

kr4g3n commented:
Is there an updated build to test?

AaronVanGeffen commented:
@marijnvdwerf Thanks for the review comments.

> Is there an updated build to test?

@kr4g3n No, there have not been any changes since the last build yet. I've been a bit too busy, unfortunately. Check back soon. :)

kr4g3n commented:
No worries, I appreciate that you're even looking into it.

I know I've asked this before, but when the game crashes, does it put a crash dump file anywhere?

AaronVanGeffen commented:
It does not, yet. You could try launching the game from a command prompt, however. Slim chance, but there might be an error printed to the console.

marijnvdwerf commented:
Transport Tycoon mobile seems to have this in a separate `Money` namespace.
```
cTTE_Handler_Money::~cTTE_Handler_Money()
cTTE_Handler_Money::~cTTE_Handler_Money()
cTTE_Handler_Money::Clear()
cTTE_Handler_Money::cTTE_Handler_Money()
cTTE_Handler_Money::cTTE_Handler_Money()
cTTE_Handler_Money::MonthlyInflation()
cTTE_Handler_Money::CalculateInflationToBaseYear(int, int, int)
cTTE_Handler_Money::GetGeneralCostReferenceWholeValue()
cTTE_Handler_Money::CalcCost(long, int, int)
cTTE_Handler_Money::CalcCostWithUnitsDistanceAndTime(long, int, int, int, int, int)
cTTE_Handler_Money::CalcCostWithUnitsDistanceAndTimeWithFractional(long, int, int, int, int, int)
cTTE_Handler_Money::GetLocalisedCurrencyValue(long, int)
cTTE_Handler_Money::GetLocalisedCurrencyValue(long long, int)
cTTE_Handler_Money::GetLocalisedCurrencyString(long, int, char*, int)
cTTE_Handler_Money::GetLocalisedCurrencyString(long long, int, char*, int)
```

marijnvdwerf commented:
Also, I think this is a good moment to add a templated `fixed_point` type.

AaronVanGeffen commented:
@marijnvdwerf I'm not sure what you're getting at with these notions of fixed points. Rather than going back and forth, would you mind implementing the desired changes yourself?

AaronVanGeffen commented:
I have rebased the PR to account for the recent wave of PRs that renamed functions to use camelCase. In the process, I have dropped @marijnvdwerf's commits that introduced a work-in-progress 'fixed point' class. This class was left unfinished, and unfortunately caused this PR to stall for more than a year.

Going forward, I hope to finish the functions can be corrected in a way that makes them correct. However, I think refactoring them to a fixed point class should really be done in another PR.

AaronVanGeffen commented on src/OpenLoco/Scenario.cpp:
```diff
@@ -12,6 +14,7 @@
 #include "Title.h"
 #include "TownManager.h"
 #include "Ui/WindowManager.h"
+#include "Ui/WindowType.h"
```
No longer needed

AaronVanGeffen commented on src/OpenLoco/Scenario.cpp:
```diff
@@ -142,9 +145,41 @@ namespace OpenLoco::Scenario
+        if ((ebx & 3) != 0)
+            eax += (ebx *= 365) + 1;
+
+        loco_global<uint32_t, 0x00525E2C> _current_day;
+        _current_day = eax;
+
+        loco_global<uint16_t, 0x00525E30> _day_progression;
+        _day_progression = 0;
+
+        loco_global<int8_t, 0x00525E34> _current_month;
+        _current_month = 0;
+
+        loco_global<int8_t, 0x00525E35> _current_day_of_month;
+        _current_day_of_month = 0;
```
Use `Date::setDate`

AaronVanGeffen commented on src/OpenLoco/Scenario.cpp:
```diff
@@ -142,9 +145,41 @@ namespace OpenLoco::Scenario
-        registers regs;
-        regs.ax = year;
-        call(0x0049685C, regs);
+        loco_global<int16_t, 0x00525E32> _current_year;
+        _current_year = year;
```
`Date::getCurrentYear`

AaronVanGeffen commented on src/OpenLoco/Economy.cpp:
```diff
@@ -1,21 +1,101 @@
+
+        // OpenLoco allows 1800 as the minimum year, whereas Locomotion uses 1900.
+        // Treat years before 1900 as though they were 1900 to not change vanilla scenarios.
+        uint32_t yearCount = std::clamp<uint32_t>(year, 1900, 2030) - 1900;
+
+        yearCount *= 12;
```
Initialise in one assignment as `monthCount`?

AaronVanGeffen commented:
The scenario stuff has been split off to #791 and merged to master, leaving only the inflation routines here. Help is welcome.

duncanspumpkin commented on src/OpenLoco/Economy.cpp:
```diff
@@ -1,21 +1,101 @@
+
+            // Storing two halves of the money struct?
+            _525EDE[i] += static_cast<uint32_t>(edx_eax);
+            currencyMultiplicationFactor[i] += static_cast<uint32_t>(edx_eax >> 32);
```
this isn't quite correct. There is a carry from 525ede.
You'll need to do:
```
uint64_t new525 = _525EDE[i] + static_cast<uint32_t>(edx_eax);
_525EDE[i] = static_cast<uint32_t>(new525);
currencyMultiplicationFactor[i] += static_cast<uint32_t>(edx_eax >> 32) + (new525 >> 32);
```

duncanspumpkin replied:
Wait, there is a much easier way of doing this its just simple maths.

```
uint64_t existingInflation = _525EDE[i] + (currencyMultiplicationFactor[i] << 32);
uint64_t additonalInflation = (existingInflation *  _4FDEC0[i]) / 4096;
uint64_t newInflation = existingInflation + additionalInflation;
_525EDE[i] = static_cast<uint32_t>(newInflation);
currencyMultiplicationFactor[i] = static_cast<uint32_t>(newInflation>> 32);
```

duncanspumpkin replied:
Oh wait but CS didn't do his calculation correctly. So this is wrong.

duncanspumpkin replied:
```
        for (uint8_t i = 0; i < 32; i++)
        {
            currencyMultiplicationFactor[i] += (static_cast<uint64_t>(_4FDEC0[i]) * currencyMultiplicationFactor[i]) >> 12;
        }
```

AaronVanGeffen replied:
Wow. I'm pleasantly surprised that chunk of code just boiled down to a oneliner in the end. Thank you very much!

duncanspumpkin commented on src/OpenLoco/Economy.cpp:
```diff
@@ -1,21 +1,101 @@
+            currencyMultiplicationFactor[i] += static_cast<uint32_t>(edx_eax);
+        }
+
+        call(0x004375F7);
```
Turn this into a sub function stub. Does some calcs on cargos.

AaronVanGeffen commented on src/OpenLoco/Economy.cpp:
```diff
@@ -1,21 +1,90 @@
 
 namespace OpenLoco::Economy
 {
+    // Inflation rates?
+    static const uint32_t _4FDEC0[32] = {
```
Shall we name this array then?
```suggestion
    static const uint32_t _inflationFactors[32] = {
```

AaronVanGeffen commented:
For what it's worth, I've played the game a few hours on this branch tonight. I'm fairly sure the inflation is working properly.
